framework/src/userobject/NearestPointLayeredAverage.C

```diff
@@ -33,3 +33,24 @@ NearestPointLayeredAverage::NearestPointLayeredAverage(const InputParameters & p
   : NearestPointBase<LayeredAverage, ElementIntegralVariableUserObject>(parameters)
 {
 }
+
+const std::vector<Point>
+NearestPointLayeredAverage::spatialPoints() const
```

Comment:
This looks like it's replicated a lot; can you share code?

Reply:
I really wasn't sure how - there's different `spatialPoints` for the `NearestPointIntegralVariablePostprocessor` and `NearestPointLayeredIntegral` for instance, but they share the same base class (the different behavior comes through the templating). So I don't think I can put `spatialPoints` in the base class. Do you have another idea?

Reply:
Can you not add a static method to some common base class with a good comment describing the behavior, and then call that static method in the virtuals? Maybe @andrsd has an idea

Reply:
Yeah I can add a method in `NearestPointBase` - it would just only be callable in certain derived classes (because the repeated code in question requires the User Objects to have `layerCenters()`, which not all of the UOs satisfy). If someone tried to use this hypothetical base class function in an incompatible derived class (one that doesn't have the `layerCenters` method), there'd be a seg fault.

I just typically stay away from cases where there's a base class method that doesn't work in all derived classes. But if that approach is acceptable to you I'm fine with it. The function would be like `NearestPointBase::spatialPointsAlongLayers`  that would be called in the relevant `spatialPoints`.

Reply:
I'll let @andrsd make a decision. I'm fine either way - but I do like seeing duplicated code reduced.

Reply:
I talked to @aprilnovak and we have a patch that should remove the code duplication at a cost of adding a dynamic_cast. That should be ok, since this would be called from `fillPoints` which hopefully is not called all the time, but only sometimes...

Reply:
Thanks for your help @andrsd :)